src/LaunchDarkly.EventSource/Internal/BufferedLineParser.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System.Threading.Tasks;
+using LaunchDarkly.EventSource.Events;
+using LaunchDarkly.EventSource.Exceptions;
+
+namespace LaunchDarkly.EventSource.Internal
+{
+    /// <summary>
+    /// Buffers a byte stream and returns it in chunks while scanning for line endings, which
+    /// may be any of CR (\r), LF (\n), or CRLF. The SSE specification allows any of these line
+    /// endings for any line in the stream.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// To use this class, we repeatedly call <see cref="ReadAsync"/> to obtain a piece of data.
+    /// Rather than copying this back into a buffer provided by the caller, BufferedLineParser
+    /// exposes its own fixed-size buffer directly and marks the portion being read; the caller
+    /// is responsible for inspecting this data before the next call to <see cref="ReadAsync"/>.
+    /// </para>
+    /// <para>
+    /// This class is not thread-safe.
+    /// </para>
+    /// </remarks>
+    internal class BufferedLineParser
```

Comment:
The logic here is a pretty straight port from [the equivalent class](https://github.com/launchdarkly/okhttp-eventsource/blob/main/src/main/java/com/launchdarkly/eventsource/BufferedLineParser.java) in `okhttp-eventsource`. The differences are basically 1. it uses async I/O and 2. it abstracts the stream with a read function rather than referencing the Stream class (see code comment immediately below).